### DISCUSSION TITLE
Apply geometric transformation to an EBSD map – is an equivalent to MATLAB function imwrap possible for EBSD data?

### DISCUSSION THREAD

**Discussion by Remsdel** [mtex-toolbox/mtex#1703] [Ask Anything]
Using porosity recognizable in an EBSD indexed map and a corresponding SEM-BSE (taken at 0 tilt) as reference points, one can define an affine geometric transformation that link the two maps. Distortions are small but not negligible.
Knowing the transformation matrix, the geometric transformation can be applied using the _imwarp_ MATLAB function (image toolbox) to transform the SEM-BSE map to match the EBSD map. The function interpolates the pixels from the SEM-BSE map so that the transformed map has also the same resolution as the EBSD map.
Now one may like to do the opposite: transform the EBSD map to match the SEM-BSE map. However the _imwarp_ function does not support EBSD data. I assume that with a combination of _affinetrans(ebsd, A)_ and _interp(ebsd,xNew,yNew)_, there might be a possibility to replicate what _imwrap_ is doing but it is not straightforward. Would somebody have the skills to adapt _imwrap_ for ebsd data or write the code that would do the same (using _affinetrans_ and _interp_ probably), it would be a useful tool to correct distortions in an EBSD map when correlated to a less-distorted image. Any advice is welcome.

**Comment by Tijmenvermeij** [comment]
Hi, you may want to check this script: https://gist.github.com/DorianDepriester/9042365805f19e8ab75cad988ff48ca6

**Reply by Remsdel** [reply]
Thanks! The script is doing something similar to the matlab function cpselect. However, it does not interpolate 'missing' data when one needs to increase the EBSD data resolution (by interpolation) to match the resolution of the corresponding SEM image.

**Comment by djm87** [comment]
I've done what you are asking using ImageJ. The following slide gives you an idea of how to do it. You'll need to interpolate the warped EBSD ID image to a grid again.
![image](https://user-images.githubusercontent.com/7084261/211415315-e7bcb525-4042-4846-a335-3506e33c1fe7.png)

**Reply by Remsdel** [reply]
Thanks! Would there be a possibility to share the ImageJ script?  Is there a reason why it was done in ImageJ and not directly in matlab?

**Reply by vtvivian** [reply]
Hi! I'm working on this exact problem at the moment... 
Could you use imwarp directly on ebsd.x and ebsd.y? 

If your warping changes the aspect ratio of ebsd.unitCell or rotates the map, you might need to call interp(ebsd,x,y) before gridify.
Or you could try gridify2 from user djm87: https://github.com/mtex-toolbox/mtex/issues/471, but it's a lot slower (written for the hexagonal case but easy to transfer across to the cubic case)

Here's an example of imwarp copied and pasted out of something I'm debugging. 

ebsd2 is the distorted data
ebsd3 is the un-distorted data
Below I've defined xy shifts (gridshift_x, gridshift_y) to use with imwarp but you could define your own affine transformation too.


[If you're interested, I can share more fully when it is eventually ready... the bug is that the unwarped xy coordinates look fine when plotting the EBSD maps, but create unexpected bugs later on in calcGrains().]



```
%% undistort EBSD data
%grids and shifts in pixel units

%we need to map TEST to REF therefore shift directions go backwards
%also need to sort out
[xGridPix, yGridPix] = meshgrid(1:NumCols,1:NumRows);
shiftXY = fliplr(cat(3,-(xGridPix-gridshift_x),(yGridPix-gridshift_y)));

%create undistorted xy positions and set all pixels outside the frame to
xGridWarp = imwarp(xGridPix, shiftXY); xGridWarp(~xGridWarp)=nan; 
yGridWarp = imwarp(yGridPix, shiftXY); yGridWarp(~yGridWarp)=nan; 

ebsd3 = EBSD(ebsd2); %un-gridify
ebsd3.x = (xGridWarp(:)*ebsd2.dx) + ebsd2.x(1,1); ebsd3.y = (yGridWarp(:)*ebsd2.dy) + ebsd2.y(1,1);
updateUnitCell(ebsd3);

ebsd3(isnan(ebsd3.x))=[]; %remove EBSD points outside the undistorted frame
[xmin2, xmax2, ymin2, ymax2] = ebsd2.extend;
ebsd3(ebsd3.x>xmax2)=[]; 
ebsd3(ebsd3.y>ymax2)=[]; 
ebsd3(ebsd3.x<xmin2)=[]; 
ebsd3(ebsd3.y<ymin2)=[]; 
```

**Comment by Remsdel** [comment]
This seems a quite promising approach. I shall try it out on my data. I would be definitely interested if you solve the issue with calcGrains().

**Comment by vtvivian** [comment]
Hi Remsdel, 

Here is a link to my code - at the moment it reads the XY pixel displacements from a .mat file from a separate tool (trueEBSD - code here: https://zenodo.org/record/4282885 ) but it should not be so hard to adapt to another kind of input for imwarp.

The xy displacements grid you read in needs to match the EBSD map coordinates of the EBSD - in my system, I think this required a manual flip. You could check whether your coordinates match up by doing something like `figure; plot(ebsd,xgrid); setMtexPref(´showCoordinates´,´on´), ` or something similar...

https://github.com/vtvivian/mtex/blob/develop/userScripts/Vivian/trueEbsd2Mtex.m

I cannot guarantee this code is correct, but I am happy to answer any questions or try to work out a solution for your use case if you would like! (and if you have any suggestions or spot any mistakes, please let me know!).

**Comment by Remsdel** [comment]
Thanks a lot for your input and precious help

I got more or less the essence of your code but still struggling to implement it for my case.

The goal is to ‘undistort’ my EBSD map by matching it to a SEM-BSE map through pore matching. The SEM-BSE map typically covers a larger area and has a higher resolution than the EBSD map (so no cropping of the EBSD should be necessary).
By matching pores I obtain a ‘deformation’ matrix tform that I hope can be used to ‘undistort’ the EBSD map onto the SEM map in same way as the imwrap function.
Basically, I would like to reproduce the output of the following function (which is not possible for an ebsd dataset unfortunately)
ebsd_undistorted=imwarp(ebsd,tform,'OutputView',imref2d(size(SEM_BSE_map)));

SEM_BSE_map, tform and ebsd are basically my only input data

I do struggle however to modify your code to replace your variables 'gridshift_x','gridshift_y','setExp','Images', phasesBsd and imgBsd
I did run in some issues running the trueEBSD – code so I could not generate those variables to better understand their role

I grasped from your code that you need to create a warped x-y grid.

So far I tried the following
xGridWarp = imwarp(xGrid_ebsd,tform,'OutputView',imref2d(size(SEM_BSE_map)));
yGridWarp = imwarp(yGrid_ebsd,tform,'OutputView',imref2d(size(SEM_BSE_map)));

but this obviously cannot work as it interpolates the grid to a wrapped grid with a higher resolution so the following will fail

ebsd3.x = (xGridWarp(:)*ebsd2.dx) + ebsd2.x(1,1); ebsd3.y = (yGridWarp(:)*ebsd2.dy) + ebsd2.y(1,1);

I think I am not fully understanding of number of your steps to get where I need. You may have a better insight on where to plug the variables SEM_BSE_map, tform and ebsd into your code. Any help would be very welcome.

**Comment by vtvivian** [comment]
Hi Remsdel, 

You can get the EBSD and BSE step sizes to match *before* calling imwarp using gridify and interp (this is my line 53).
I don't really crop any of the images either - just the 'overhang' bits where not all the pixel numbers are exact multiples of each other, and much later on when I want to make my undistorted map a rectangle again.

`%resize EBSD map
ebsd2 = gridify(interp(ebsd1, xGrid(:), yGrid(:))); `

xGrid and yGrid can be produced by calling meshgrid on your SEM_BSE_map coordinate vectors (you'll also need the BSE step size, in my function it's setExp.PixSizeImg). (This is line 23-50 in my code - it gets a bit complicated because you need to deal with step size rounding issues when going between the BSE and EBSD data, and the fact that the 'zero' xy position is not always the top left pixel anymore if you increase the step size!)

**Comment by Remsdel** [comment]
Ok thanks that clarifies a lot the first step. Just to make sure: to what corresponds the ‘images’ variable in your case?
________________________________

**Reply by vtvivian** [reply]
Images is a structure which contains all the images in its fields (so Images.BSE is the BSE image, which is a 1024*768 double matrix or similar)
(Using structures in this way was perhaps not the best way to organise the data - sorry)

**Reply by Fred-Ad** [reply]
Hi everyone, 

For this kind f problem I had to make a proper stitching of low mag EBSD map, I used this kind of code 

```matlab
%%Measure the transform matrix
fixed = imread("C:\Users\d597668\Documents\Prestas\2021-094\Exploitation\fixed.jpg");
moving = imread("C:\Users\d597668\Documents\Prestas\2021-094\Exploitation\moving.jpg");
cpselect(moving,fixed);
fixedPoints(:,2)=max(fixedPoints(:,2))-fixedPoints(:,2)
movingPoints(:,2)=max(movingPoints(:,2))-movingPoints(:,2)
tform = fitgeotrans(movingPoints,fixedPoints,'projective');
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%Apply the transform matrix to ebsd coordinates in Mtex
ebsd.x=ebsd.x-min(ebsd.x);
ebsd.y=ebsd.y-min(ebsd.y);
U=ebsd.x;
V=ebsd.y;
[Y,X] = transformPointsInverse(tform,V,U);
ebsd.x=X;ebsd.y=Y;
```

4 points : 
- I used a simple projective transform, I think it can work with any transform matrix.
- You have to be careful about X,Y and U,V to match your EBSD map ref and image ref --> you need a ref coord (0,0) at the same corner of your map and ref image.
- You need the same number of px on both images to correctly calculate tform
- You can gridify the result with the original unitCell to have a rectangular EBSD map at least

It worked for me.

Regards
